File: compiler/ast.py
"""AST node definitions for Droe DSL compiler."""

from dataclasses import dataclass, field
from typing import Any, List, Optional, Union


@dataclass
class ASTNode:
    """Base class for all AST nodes."""
    line_number: Optional[int] = field(default=None, init=False)


@dataclass
class Literal(ASTNode):
    """Represents a literal value (string, number, boolean)."""
    value: Union[str, int, float, bool]
    type: str  # 'string', 'number', 'boolean'


@dataclass
class Identifier(ASTNode):
    """Represents an identifier (variable name)."""
    name: str


@dataclass
class BinaryOp(ASTNode):
    """Represents a binary operation (e.g., >, <, ==, +, -)."""
    left: ASTNode
    operator: str
    right: ASTNode


@dataclass
class DisplayStatement(ASTNode):
    """Represents a display statement."""
    expression: ASTNode


@dataclass
class IfStatement(ASTNode):
    """Represents an if-then statement."""
    condition: ASTNode
    then_body: List[ASTNode]
    else_body: Optional[List[ASTNode]] = None


@dataclass
class PropertyAccess(ASTNode):
    """Represents property access (e.g., user.age)."""
    object: ASTNode
    property: str


@dataclass
class Assignment(ASTNode):
    """Represents a variable assignment (set x to value)."""
    variable: str
    value: ASTNode


@dataclass
class ArrayLiteral(ASTNode):
    """Represents an array literal like ["a", "b", "c"]."""
    elements: List[ASTNode]


@dataclass
class WhileLoop(ASTNode):
    """Represents a while loop."""
    condition: ASTNode
    body: List[ASTNode]


@dataclass
class ForEachLoop(ASTNode):
    """Represents a for each loop."""
    variable: str
    iterable: ASTNode
    body: List[ASTNode]


@dataclass
class ArithmeticOp(ASTNode):
    """Represents arithmetic operations (+, -, *, /)."""
    left: ASTNode
    operator: str
    right: ASTNode


@dataclass
class TaskAction(ASTNode):
    """Represents a task action definition (task name with params ... end)."""
    name: str
    parameters: List['ActionParameter'] = field(default_factory=list)  # Reuse ActionParameter for consistency
    body: List[ASTNode] = field(default_factory=list)


@dataclass
class TaskInvocation(ASTNode):
    """Represents a task invocation (run task_name with args)."""
    task_name: str
    arguments: List[ASTNode] = field(default_factory=list)  # Arguments for parameterized tasks


@dataclass
class ActionDefinition(ASTNode):
    """Represents an action definition (action name ... end action)."""
    name: str
    body: List[ASTNode]


@dataclass
class ReturnStatement(ASTNode):
    """Represents a return statement (respond with, answer is, output, give)."""
    expression: ASTNode
    return_type: str  # 'respond_with', 'answer_is', 'output', 'give'


@dataclass
class ActionInvocation(ASTNode):
    """Represents an action invocation that returns a value."""
    action_name: str


@dataclass
class ModuleDefinition(ASTNode):
    """Represents a module definition (module name ... end module)."""
    name: str
    body: List[ASTNode]


@dataclass
class DataDefinition(ASTNode):
    """Represents a data structure definition (data Name ... end data)."""
    name: str
    fields: List['DataField']
    storage_type: Optional[str] = None  # 'short_store' (sessionStorage) or 'long_store' (localStorage)


@dataclass
class DataField(ASTNode):
    """Represents a field in a data structure."""
    name: str
    type: str
    annotations: List[str] = field(default_factory=list)  # e.g., ['required', 'unique', 'key', 'auto']


@dataclass
class ActionDefinitionWithParams(ASTNode):
    """Represents an action definition with parameters and return type."""
    name: str
    parameters: List['ActionParameter']
    return_type: Optional[str]
    body: List[ASTNode]


@dataclass
class ActionParameter(ASTNode):
    """Represents a parameter in an action definition."""
    name: str
    type: str


@dataclass
class ActionInvocationWithArgs(ASTNode):
    """Represents an action invocation with arguments."""
    module_name: Optional[str]  # For module.action calls
    action_name: str
    arguments: List[ASTNode]


@dataclass
class StringInterpolation(ASTNode):
    """Represents a string with variable interpolation like 'Hello [name]'."""
    parts: List[ASTNode]  # Mix of Literal (for text) and Identifier (for variables)


@dataclass
class DataInstance(ASTNode):
    """Represents a data structure instance creation."""
    data_type: str  # The data type name (e.g., "User")
    field_values: List['FieldAssignment']  # Field assignments


@dataclass
class FieldAssignment(ASTNode):
    """Represents a field assignment in data instance creation."""
    field_name: str
    value: ASTNode


@dataclass
class IncludeStatement(ASTNode):
    """Represents an include statement (include ModuleName.droe)."""
    module_name: str  # The module name without .droe extension
    file_path: str    # The full file path (ModuleName.droe)


@dataclass
class AssetInclude(ASTNode):
    """Represents an asset include statement (include assets/style.css)."""
    asset_path: str  # Path to asset file
    asset_type: str  # 'css', 'js', 'font', etc.


@dataclass
class FormatExpression(ASTNode):
    """Represents a format expression (format variable as "pattern")."""
    expression: ASTNode  # The expression to format
    format_pattern: str  # The format pattern string


@dataclass
class MetadataAnnotation(ASTNode):
    """Represents a metadata annotation like @target web or @metadata(platform="mobile")."""
    key: str  # The annotation key (target, name, description, metadata)
    value: Union[str, dict]  # The annotation value or parameters dict


@dataclass
class FragmentDefinition(ASTNode):
    """Represents a fragment definition (fragment name ... end fragment)."""
    name: str
    slots: List['SlotComponent'] = field(default_factory=list)
    attributes: List['AttributeDefinition'] = field(default_factory=list)
    classes: List[str] = field(default_factory=list)  # CSS classes
    styles: Optional[str] = None  # Inline styles


@dataclass
class FragmentReference(ASTNode):
    """Represents a reference to a fragment with slot content (fragment name ... end fragment)."""
    fragment_name: str
    slot_contents: dict = field(default_factory=dict)  # Map slot names to content lists


@dataclass
class ScreenDefinition(ASTNode):
    """Represents a screen definition (screen name ... end screen)."""
    name: str
    fragments: List['FragmentReference'] = field(default_factory=list)  # Fragments used in this screen
    attributes: List['AttributeDefinition'] = field(default_factory=list)
    classes: List[str] = field(default_factory=list)  # CSS classes
    styles: Optional[str] = None  # Inline styles


@dataclass
class SlotComponent(ASTNode):
    """Represents a slot component for fragments (slot "name")."""
    name: str  # Slot name like "content", "header", "footer", "sidebar"
    default_content: List[ASTNode] = field(default_factory=list)  # Optional default content
    attributes: List['AttributeDefinition'] = field(default_factory=list)
    classes: List[str] = field(default_factory=list)  # CSS classes
    styles: Optional[str] = None  # Inline styles
    component_type: str = field(default="slot", init=False)


@dataclass
class FormDefinition(ASTNode):
    """Represents a form definition (form name ... end form)."""
    name: str
    children: List[ASTNode]
    attributes: List['AttributeDefinition'] = field(default_factory=list)
    classes: List[str] = field(default_factory=list)  # CSS classes
    styles: Optional[str] = None  # Inline styles


@dataclass
class TitleComponent(ASTNode):
    """Represents a title component."""
    text: str
    attributes: List['AttributeDefinition'] = field(default_factory=list)
    classes: List[str] = field(default_factory=list)  # CSS classes
    styles: Optional[str] = None  # Inline styles
    component_type: str = field(default="title", init=False)


@dataclass
class TextComponent(ASTNode):
    """Represents a text component for displaying text content."""
    text: str
    attributes: List['AttributeDefinition'] = field(default_factory=list)
    classes: List[str] = field(default_factory=list)  # CSS classes
    styles: Optional[str] = None  # Inline styles
    component_type: str = field(default="text", init=False)


@dataclass
class InputComponent(ASTNode):
    """Represents an input component."""
    input_type: str = "text"  # 'text', 'password', 'email', etc.
    binding: Optional[str] = None  # Data binding target
    attributes: List['AttributeDefinition'] = field(default_factory=list)
    element_id: Optional[str] = None  # Element ID for form handling
    classes: List[str] = field(default_factory=list)  # CSS classes
    styles: Optional[str] = None  # Inline styles
    component_type: str = field(default="input", init=False)


@dataclass
class TextareaComponent(ASTNode):
    """Represents a textarea component."""
    label: Optional[str] = None
    placeholder: Optional[str] = None
    rows: Optional[int] = None
    binding: Optional[str] = None
    attributes: List['AttributeDefinition'] = field(default_factory=list)
    classes: List[str] = field(default_factory=list)  # CSS classes
    styles: Optional[str] = None  # Inline styles
    element_id: Optional[str] = None  # Element ID for form handling
    component_type: str = field(default="textarea", init=False)


@dataclass
class DropdownComponent(ASTNode):
    """Represents a dropdown/select component."""
    label: Optional[str] = None
    options: List[ASTNode] = field(default_factory=list)
    binding: Optional[str] = None
    attributes: List['AttributeDefinition'] = field(default_factory=list)
    element_id: Optional[str] = None  # Element ID for form handling
    classes: List[str] = field(default_factory=list)  # CSS classes
    styles: Optional[str] = None  # Inline styles
    component_type: str = field(default="dropdown", init=False)


@dataclass
class ToggleComponent(ASTNode):
    """Represents a toggle/switch component."""
    binding: Optional[str] = None
    attributes: List['AttributeDefinition'] = field(default_factory=list)
    classes: List[str] = field(default_factory=list)  # CSS classes
    styles: Optional[str] = None  # Inline styles
    element_id: Optional[str] = None  # Element ID for form handling
    component_type: str = field(default="toggle", init=False)


@dataclass
class CheckboxComponent(ASTNode):
    """Represents a checkbox component."""
    text: Optional[str] = None
    binding: Optional[str] = None
    attributes: List['AttributeDefinition'] = field(default_factory=list)
    element_id: Optional[str] = None  # Element ID for form handling
    classes: List[str] = field(default_factory=list)  # CSS classes
    styles: Optional[str] = None  # Inline styles
    component_type: str = field(default="checkbox", init=False)


@dataclass
class RadioComponent(ASTNode):
    """Represents a radio button component."""
    text: Optional[str] = None
    value: Optional[str] = None
    binding: Optional[str] = None
    attributes: List['AttributeDefinition'] = field(default_factory=list)
    element_id: Optional[str] = None  # Element ID for form handling
    classes: List[str] = field(default_factory=list)  # CSS classes
    styles: Optional[str] = None  # Inline styles
    component_type: str = field(default="radio", init=False)


@dataclass
class ButtonComponent(ASTNode):
    """Represents a button component."""
    text: str
    action: Optional[str] = None  # Action to run on click
    attributes: List['AttributeDefinition'] = field(default_factory=list)
    classes: List[str] = field(default_factory=list)  # CSS classes
    styles: Optional[str] = None  # Inline styles
    component_type: str = field(default="button", init=False)


@dataclass
class ImageComponent(ASTNode):
    """Represents an image component."""
    src: str  # Image source path
    alt: Optional[str] = None  # Alt text
    attributes: List['AttributeDefinition'] = field(default_factory=list)
    classes: List[str] = field(default_factory=list)  # CSS classes
    styles: Optional[str] = None  # Inline styles
    component_type: str = field(default="image", init=False)


@dataclass
class VideoComponent(ASTNode):
    """Represents a video component."""
    src: str  # Video source path
    controls: bool = True
    autoplay: bool = False
    loop: bool = False
    muted: bool = False
    attributes: List['AttributeDefinition'] = field(default_factory=list)
    classes: List[str] = field(default_factory=list)  # CSS classes
    styles: Optional[str] = None  # Inline styles
    component_type: str = field(default="video", init=False)


@dataclass
class AudioComponent(ASTNode):
    """Represents an audio component."""
    src: str  # Audio source path
    controls: bool = True
    autoplay: bool = False
    loop: bool = False
    attributes: List['AttributeDefinition'] = field(default_factory=list)
    classes: List[str] = field(default_factory=list)  # CSS classes
    styles: Optional[str] = None  # Inline styles
    component_type: str = field(default="audio", init=False)


@dataclass
class AttributeDefinition(ASTNode):
    """Represents an attribute definition (validate required, bind LoginForm.email, etc.)."""
    name: str
    value: Optional[ASTNode] = None  # Can be a literal or expression


@dataclass
class ValidationAttribute(ASTNode):
    """Represents a validation attribute."""
    validation_type: str  # 'required', 'email', 'numeric', etc.
    name: str = field(default="validate", init=False)


@dataclass
class BindingAttribute(ASTNode):
    """Represents a data binding attribute."""
    binding_target: str  # e.g., 'LoginForm.email'
    name: str = field(default="bind", init=False)


@dataclass
class ActionAttribute(ASTNode):
    """Represents an action attribute (run action_name)."""
    action_name: str
    name: str = field(default="run", init=False)


@dataclass
class ApiCallStatement(ASTNode):
    """Represents an API call statement (call/fetch /endpoint method GET/POST with data)."""
    verb: str  # 'call', 'fetch', 'update', 'delete'
    endpoint: str  # '/login', '/user/profile', etc.
    method: str  # 'GET', 'POST', 'PUT', 'DELETE'
    payload: Optional[str] = None  # Variable name for request body (e.g., 'loginForm')
    headers: List['ApiHeader'] = field(default_factory=list)  # List of headers
    response_variable: Optional[str] = None  # Variable to store response (e.g., 'response')
    
    
@dataclass
class ApiHeader(ASTNode):
    """Represents an API header (Authorization: "Bearer Token")."""
    name: str  # 'Authorization', 'Content-Type', etc.
    value: str  # '"Bearer Token"', '"application/json"', etc.




@dataclass
class ServeStatement(ASTNode):
    """Represents a serve statement (serve get/post/put/delete /endpoint ... end serve)."""
    method: str  # 'get', 'post', 'put', 'delete'
    endpoint: str  # '/user', '/user/:id', etc.
    body: List[ASTNode]  # Statements inside the serve block
    params: List[str] = field(default_factory=list)  # URL parameters
    accept_type: Optional[str] = None  # Type for request body (accept statement)
    response_action: Optional[ASTNode] = None  # Action called for response

@dataclass
class AcceptStatement(ASTNode):
    """Represents an accept statement (accept TypeName.actionName)."""
    module_name: str
    action_name: str
    param_name: Optional[str] = None  # For accept ... with param_name

@dataclass
class RespondStatement(ASTNode):
    """Represents a respond statement (respond with ModuleName.actionName)."""
    module_name: str
    action_name: str
    param_name: Optional[str] = None  # For respond with ... with param_name

@dataclass
class ParamsStatement(ASTNode):
    """Represents a params statement (params id which is text)."""
    param_name: str
    param_type: str

@dataclass
class DatabaseStatement(ASTNode):
    """Represents a database operation (db find/create/update User ...)."""
    operation: str  # 'find', 'create', 'update', 'delete'
    entity_name: str  # User, Product, etc.
    conditions: List[ASTNode] = field(default_factory=list)  # where clauses
    fields: List[ASTNode] = field(default_factory=list)  # set clauses for update
    return_var: Optional[str] = None  # return id into variable

@dataclass
class Program(ASTNode):
    """Root node containing all statements in the program."""
    statements: List[ASTNode]
    metadata: List[MetadataAnnotation] = field(default_factory=list)  # Metadata annotations
    included_modules: List['IncludeStatement'] = None  # Track included modules

File: droe-compiler/src/ast.rs
//! AST node definitions for Droe DSL compiler.

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Base trait for all AST nodes
pub trait ASTNode {
    fn line_number(&self) -> Option<usize>;
    fn set_line_number(&mut self, line: usize);
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum Node {
    // Literals and identifiers
    Literal(Literal),
    Identifier(Identifier),
    
    // Expressions
    BinaryOp(BinaryOp),
    PropertyAccess(PropertyAccess),
    ArrayLiteral(ArrayLiteral),
    ArithmeticOp(ArithmeticOp),
    StringInterpolation(StringInterpolation),
    FormatExpression(FormatExpression),
    
    // Statements
    DisplayStatement(DisplayStatement),
    Assignment(Assignment),
    IfStatement(IfStatement),
    WhileLoop(WhileLoop),
    ForEachLoop(ForEachLoop),
    ReturnStatement(ReturnStatement),
    
    // Definitions
    ModuleDefinition(ModuleDefinition),
    DataDefinition(DataDefinition),
    ActionDefinition(ActionDefinition),
    ActionDefinitionWithParams(ActionDefinitionWithParams),
    TaskAction(TaskAction),
    FragmentDefinition(FragmentDefinition),
    ScreenDefinition(ScreenDefinition),
    FormDefinition(FormDefinition),
    
    // Components
    TitleComponent(TitleComponent),
    TextComponent(TextComponent),
    InputComponent(InputComponent),
    TextareaComponent(TextareaComponent),
    DropdownComponent(DropdownComponent),
    ToggleComponent(ToggleComponent),
    CheckboxComponent(CheckboxComponent),
    RadioComponent(RadioComponent),
    ButtonComponent(ButtonComponent),
    ImageComponent(ImageComponent),
    VideoComponent(VideoComponent),
    AudioComponent(AudioComponent),
    SlotComponent(SlotComponent),
    
    // API and Database
    ApiCallStatement(ApiCallStatement),
    DatabaseStatement(DatabaseStatement),
    ServeStatement(ServeStatement),
    
    // Includes and Metadata
    IncludeStatement(IncludeStatement),
    AssetInclude(AssetInclude),
    MetadataAnnotation(MetadataAnnotation),
    
    // Other
    TaskInvocation(TaskInvocation),
    ActionInvocation(ActionInvocation),
    ActionInvocationWithArgs(ActionInvocationWithArgs),
    DataInstance(DataInstance),
    FieldAssignment(FieldAssignment),
    AcceptStatement(AcceptStatement),
    RespondStatement(RespondStatement),
    ParamsStatement(ParamsStatement),
    Program(Program),
}

impl ASTNode for Node {
    fn line_number(&self) -> Option<usize> {
        match self {
            Node::Literal(n) => n.line_number,
            Node::Identifier(n) => n.line_number,
            Node::BinaryOp(n) => n.line_number,
            // Add for all variants...
            _ => None, // TODO: implement for all variants
        }
    }
    
    fn set_line_number(&mut self, line: usize) {
        match self {
            Node::Literal(n) => n.line_number = Some(line),
            Node::Identifier(n) => n.line_number = Some(line),
            Node::BinaryOp(n) => n.line_number = Some(line),
            // Add for all variants...
            _ => {} // TODO: implement for all variants
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Literal {
    pub value: LiteralValue,
    pub literal_type: String,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum LiteralValue {
    String(String),
    Integer(i64),
    Float(f64),
    Boolean(bool),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Identifier {
    pub name: String,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BinaryOp {
    pub left: Box<Node>,
    pub operator: String,
    pub right: Box<Node>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DisplayStatement {
    pub expression: Box<Node>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct IfStatement {
    pub condition: Box<Node>,
    pub then_body: Vec<Node>,
    pub else_body: Option<Vec<Node>>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PropertyAccess {
    pub object: Box<Node>,
    pub property: String,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Assignment {
    pub variable: String,
    pub value: Box<Node>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ArrayLiteral {
    pub elements: Vec<Node>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WhileLoop {
    pub condition: Box<Node>,
    pub body: Vec<Node>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ForEachLoop {
    pub variable: String,
    pub iterable: Box<Node>,
    pub body: Vec<Node>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ArithmeticOp {
    pub left: Box<Node>,
    pub operator: String,
    pub right: Box<Node>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TaskAction {
    pub name: String,
    pub parameters: Vec<ActionParameter>,
    pub body: Vec<Node>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TaskInvocation {
    pub task_name: String,
    pub arguments: Vec<Node>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ActionDefinition {
    pub name: String,
    pub body: Vec<Node>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ReturnStatement {
    pub expression: Box<Node>,
    pub return_type: String,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ActionInvocation {
    pub action_name: String,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ModuleDefinition {
    pub name: String,
    pub body: Vec<Node>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DataDefinition {
    pub name: String,
    pub fields: Vec<DataField>,
    pub storage_type: Option<String>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DataField {
    pub name: String,
    pub field_type: String,
    pub annotations: Vec<String>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ActionDefinitionWithParams {
    pub name: String,
    pub parameters: Vec<ActionParameter>,
    pub return_type: Option<String>,
    pub body: Vec<Node>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ActionParameter {
    pub name: String,
    pub param_type: String,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ActionInvocationWithArgs {
    pub module_name: Option<String>,
    pub action_name: String,
    pub arguments: Vec<Node>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StringInterpolation {
    pub parts: Vec<Node>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DataInstance {
    pub data_type: String,
    pub field_values: Vec<FieldAssignment>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FieldAssignment {
    pub field_name: String,
    pub value: Box<Node>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct IncludeStatement {
    pub module_name: String,
    pub file_path: String,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AssetInclude {
    pub asset_path: String,
    pub asset_type: String,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FormatExpression {
    pub expression: Box<Node>,
    pub format_pattern: String,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MetadataAnnotation {
    pub key: String,
    pub value: MetadataValue,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum MetadataValue {
    String(String),
    Dict(HashMap<String, String>),
}

// UI Components
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FragmentDefinition {
    pub name: String,
    pub slots: Vec<SlotComponent>,
    pub attributes: Vec<AttributeDefinition>,
    pub classes: Vec<String>,
    pub styles: Option<String>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ScreenDefinition {
    pub name: String,
    pub fragments: Vec<FragmentReference>,
    pub attributes: Vec<AttributeDefinition>,
    pub classes: Vec<String>,
    pub styles: Option<String>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FragmentReference {
    pub fragment_name: String,
    pub slot_contents: HashMap<String, Vec<Node>>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SlotComponent {
    pub name: String,
    pub default_content: Vec<Node>,
    pub attributes: Vec<AttributeDefinition>,
    pub classes: Vec<String>,
    pub styles: Option<String>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FormDefinition {
    pub name: String,
    pub children: Vec<Node>,
    pub attributes: Vec<AttributeDefinition>,
    pub classes: Vec<String>,
    pub styles: Option<String>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TitleComponent {
    pub text: String,
    pub attributes: Vec<AttributeDefinition>,
    pub classes: Vec<String>,
    pub styles: Option<String>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TextComponent {
    pub text: String,
    pub attributes: Vec<AttributeDefinition>,
    pub classes: Vec<String>,
    pub styles: Option<String>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InputComponent {
    pub input_type: String,
    pub binding: Option<String>,
    pub attributes: Vec<AttributeDefinition>,
    pub element_id: Option<String>,
    pub classes: Vec<String>,
    pub styles: Option<String>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TextareaComponent {
    pub label: Option<String>,
    pub placeholder: Option<String>,
    pub rows: Option<i32>,
    pub binding: Option<String>,
    pub attributes: Vec<AttributeDefinition>,
    pub classes: Vec<String>,
    pub styles: Option<String>,
    pub element_id: Option<String>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DropdownComponent {
    pub label: Option<String>,
    pub options: Vec<Node>,
    pub binding: Option<String>,
    pub attributes: Vec<AttributeDefinition>,
    pub element_id: Option<String>,
    pub classes: Vec<String>,
    pub styles: Option<String>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToggleComponent {
    pub binding: Option<String>,
    pub attributes: Vec<AttributeDefinition>,
    pub classes: Vec<String>,
    pub styles: Option<String>,
    pub element_id: Option<String>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CheckboxComponent {
    pub text: Option<String>,
    pub binding: Option<String>,
    pub attributes: Vec<AttributeDefinition>,
    pub element_id: Option<String>,
    pub classes: Vec<String>,
    pub styles: Option<String>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RadioComponent {
    pub text: Option<String>,
    pub value: Option<String>,
    pub binding: Option<String>,
    pub attributes: Vec<AttributeDefinition>,
    pub element_id: Option<String>,
    pub classes: Vec<String>,
    pub styles: Option<String>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ButtonComponent {
    pub text: String,
    pub action: Option<String>,
    pub attributes: Vec<AttributeDefinition>,
    pub classes: Vec<String>,
    pub styles: Option<String>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ImageComponent {
    pub src: String,
    pub alt: Option<String>,
    pub attributes: Vec<AttributeDefinition>,
    pub classes: Vec<String>,
    pub styles: Option<String>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VideoComponent {
    pub src: String,
    pub controls: bool,
    pub autoplay: bool,
    pub loop_video: bool,
    pub muted: bool,
    pub attributes: Vec<AttributeDefinition>,
    pub classes: Vec<String>,
    pub styles: Option<String>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AudioComponent {
    pub src: String,
    pub controls: bool,
    pub autoplay: bool,
    pub loop_audio: bool,
    pub attributes: Vec<AttributeDefinition>,
    pub classes: Vec<String>,
    pub styles: Option<String>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AttributeDefinition {
    pub name: String,
    pub value: Option<Box<Node>>,
    pub line_number: Option<usize>,
}

// API and Database
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ApiCallStatement {
    pub verb: String,
    pub endpoint: String,
    pub method: String,
    pub payload: Option<String>,
    pub headers: Vec<ApiHeader>,
    pub response_variable: Option<String>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ApiHeader {
    pub name: String,
    pub value: String,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ServeStatement {
    pub method: String,
    pub endpoint: String,
    pub body: Vec<Node>,
    pub params: Vec<String>,
    pub accept_type: Option<String>,
    pub response_action: Option<Box<Node>>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AcceptStatement {
    pub module_name: String,
    pub action_name: String,
    pub param_name: Option<String>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RespondStatement {
    pub module_name: String,
    pub action_name: String,
    pub param_name: Option<String>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ParamsStatement {
    pub param_name: String,
    pub param_type: String,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DatabaseStatement {
    pub operation: String,
    pub entity_name: String,
    pub conditions: Vec<Node>,
    pub fields: Vec<Node>,
    pub return_var: Option<String>,
    pub line_number: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Program {
    pub statements: Vec<Node>,
    pub metadata: Vec<MetadataAnnotation>,
    pub included_modules: Option<Vec<IncludeStatement>>,
    pub line_number: Option<usize>,
}

// Error types
#[derive(Debug, Clone, Serialize, Deserialize, thiserror::Error)]
#[error("{message} at line {line}, column {column}")]
pub struct ParseError {
    pub message: String,
    pub line: usize,
    pub column: usize,
}

pub type ParseResult<T> = Result<T, ParseError>;